Bootstrap test suite for front-end app

mongrelion opened the pull request:
Things that you will find in this pull request:
- Small refactor of the front-end app: I moved the routes definitions out of the `server.js` file into their own modules to make adding tests more easy. Tests are WIP
- Setup the test suite for unit, functional and end-to-end tests: Chai, Mocha, Phantomjs, Casperjs, Makefile with tasks for running the tests, preparing containers, etc.
- Added some basic unit tests for the API
- Added some basic e2e tests for testing the following flows: home page, login, cart interaction, checkout.

Read the README file for instructions on how to run the tests.

Commit:
Prepare front-end app for testing

* Added the following libraries:
  - mocha for running test
  - chai for assertions/expectations
  - istanbul for code coverage
* Added a Makefile to make development easier by defining tasks for
  - installing dependencies
  - running the web server
  - running tests
  - generating test coverage reports
* Updated authors/contributors
* Fixed url to project source in package.json file

Commit:
Run tests with style

Commit:
Fix formatting

2 spaces instead of 4. No pythonist.

Commit:
Move helpers out into its own lib

Commit:
Update Makefile with dev goodies

Added tasks for:
* Creating the Docker image for development
* Running the container and attaching it to the proper Docker networks
* Deleting old running instances
* Running tests in watch mode (check mocha -w)
* Follow-tailng the container's logs
* Opening the browser in the shop

Commit:
Extract routes into their own modules

Commit:
Move config and error handling out of server

Commit:
Fix test runner

* Recursive mode for mocha
* Fix browser task

Commit:
Simplify Dockerfile

Commit:
Add steps for running end-2-end tests

Commit:
Add integration tests for home page and login scenarios

Commit:
Add integration tests for the catalogue

Commit:
Add integration tests for the cart

Commit:
Add integration tests for checkout

Commit:
Add script for running e2e tests

Currently the script cleans the database up and runs each test file
afterwards.

Commit:
Update README file

sashkachan commented on sockshop/front-end/Makefile:
```diff
@@ -0,0 +1,37 @@
+.PHONY: test coverage
```
Don't you need more PHONY targets in here?

mongrelion replied:
Not really. I tried every single task in the Makefile (I wrote it myself) and the only two tasks that have matching folders are these two.

Commit:
Fix `test` task in Makefile

Passing the --recursive option to mocha was breaking the test suite as
it was trying to load as well the integration specs, which require
casper which is not available for the unit tests.
